File: src/c-call-f/c-call-f.c
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>

void print_1d_f(int n, int *a);

void print_2d_f_right1(int n0, int n1, int a[n1][n0]);

void print_2d_f_right2(int n0, int n1, int *a);

void print_2d_f_wrong(int n0, int n1, int **a);

int main() {
  int i;
  int j;

  // 1d array
  int n = 3;
  int *b;

  b = (int *) malloc(sizeof(int) * n);
  for (i = 0; i < n; i++)
    b[i] = i;
  printf("print 1d array\n");

  // int n = 3;
  // int *b;
  print_1d_f(n, b);

  // 2d array
  int n0 = 2;
  int n1 = 3;
  int c[n1][n0];
  int **a;

  a = (int**) malloc(sizeof(int*) * n1);
  for (i = 0; i < n1; i++) {
    a[i] = (int*) malloc(sizeof(int) * n0);
    for (j = 0; j < n0; j++) {
      a[i][j] = i+j;
      c[i][j] = i+j;
    }
  }

  printf("the values are wrong\n");
  printf("print 2d array\n");
  print_2d_f_wrong(n0, n1, a);

  printf("the values are right1\n");
  // int n0 = 2;
  // int n1 = 3;
  // int c[n1][n0];
  print_2d_f_right1(n0, n1, c);

  printf("the values are right2\n");
  print_2d_f_right2(n0, n1, &c[0][0]);
  return 0;
}

File: src/c-call-f/c-call-f.f90
! -*- coding: utf-8 -*-
!===============================================================================
!> @brief
!> @author Qingming He, NECP
!> @date 2019-07-27 11:31:11
!===============================================================================
module c_call_f

  use iso_c_binding

  implicit none

contains

  subroutine print_1d(a)
    integer :: a(:)
    integer :: i
    do i = 1, size(a)
      write(*, *) a(i)
    end do
  end subroutine print_1d

  subroutine print_2d(a)
    integer :: a(:, :)
    integer :: n0
    integer :: n1
    integer :: i
    integer :: j
    n0 = size(a, 1)
    n1 = size(a, 2)
    do j = 1, n1
      do i = 1, n0
        write(*, *) a(i, j)
      end do
    end do
  end subroutine print_2d

  subroutine print_1d_f(n, a) bind(C, name="print_1d_f")
    integer(C_INT), value :: n
    integer(C_INT) :: a(n)
    call print_1d(a)
  end subroutine print_1d_f

  subroutine print_2d_f_right1(n0, n1, a) bind(C, name="print_2d_f_right1")
    integer(C_INT), value :: n0
    integer(C_INT), value :: n1
    integer(C_INT) :: a(n0, n1)
    call print_2d(a)
  end subroutine print_2d_f_right1

  subroutine print_2d_f_right2(n0, n1, a) bind(C, name="print_2d_f_right2")
    integer(C_INT), value :: n0
    integer(C_INT), value :: n1
    integer(C_INT) :: a(n0, n1)

    call print_2d(a)

  end subroutine print_2d_f_right2

  subroutine print_2d_f_wrong(n0, n1, a) bind(C, name="print_2d_f_wrong")
    integer(C_INT), value :: n0
    integer(C_INT), value :: n1
    integer(C_INT) :: a(n0, n1)

    call print_2d(a)

  end subroutine print_2d_f_wrong

end module c_call_f
